reporter/gcppubsub/pubsub.go

```diff
@@ -0,0 +1,109 @@
+package gcppubsub
+
+import (
+	"cloud.google.com/go/pubsub"
+	"context"
+	"encoding/json"
+	"github.com/openzipkin/zipkin-go/model"
+	"github.com/openzipkin/zipkin-go/reporter"
+	"log"
+	"os"
+)
+
+const defaultPubSubTopic = "defaultTopic"
+
+// Reporter implements Reporter by publishing spans to a GCP gcppubsub.
+type Reporter struct {
+	logger *log.Logger
+	topic  string
+	client *pubsub.Client
+}
+
+// ReporterOption sets a parameter for the PubSubReporter
+type ReporterOption func(c *Reporter)
+
+// Send send span to topic
+func (r *Reporter) Send(s model.SpanModel) {
+	// Zipkin expects the message to be wrapped in an array
+	ss := []model.SpanModel{s}
+	m, err := json.Marshal(ss)
+	if err != nil {
+		r.logger.Printf("failed when marshalling the span: %s\n", err.Error())
+		return
+	}
+	err = r.publish(m)
+	if err != nil {
+		r.logger.Printf("Error publishing message to gcppubsub: %s msg: %s", err.Error(), string(m))
+	}
+}
+
+// Close close span
+func (r *Reporter) Close() error {
+	return r.client.Close()
+}
+
+// Logger sets the logger used to report errors in the collection
+// process.
+func Logger(logger *log.Logger) ReporterOption {
+	return func(c *Reporter) {
+		c.logger = logger
+	}
+}
+
+// Client sets the client used to produce to gcppubsub.
+func Client(client *pubsub.Client) ReporterOption {
+	return func(c *Reporter) {
+		c.client = client
+	}
+}
+
+// Topic sets the gcppubsub topic to attach the reporter producer on.
+func Topic(t string) ReporterOption {
+	return func(c *Reporter) {
+		c.topic = t
+	}
+}
+
+// NewReporter returns a new gcppubsub-backed Reporter. address should be a slice of
+// TCP endpoints of the form "host:port".
+func NewReporter(options ...ReporterOption) (reporter.Reporter, error) {
+	topic := os.Getenv("ZIPKIN_PUBSUB_TOPIC")
+	if topic == "" {
+		topic = defaultPubSubTopic
+	}
+
+	r := &Reporter{
+		logger: log.New(os.Stderr, "", log.LstdFlags),
+		topic:  defaultPubSubTopic,
+	}
+
+	for _, option := range options {
+		option(r)
+	}
+	if r.client == nil {
+		ctx := context.Background()
+		proj := os.Getenv("GOOGLE_CLOUD_PROJECT")
```

Comment:
I have mixed feelings here. In one hand I usually don't like env vars support for libraries but I don't see a trivial way to not to have mutually exclusive options like client and google_cloud_project. Maybe the env var is good enough. Any thoughts @basvanbeek ?

Reply:
There is a trivial one... requiring a valid pubsub client as input parameter to NewReporter instead of as a functional option... 

Thinking about it I like the idea of requiring this to be explicit...